Make user objects with spatialValue implementations functors

lindsayad opened the pull request:
Closes #29910

lindsayad committed:
Make `LayeredIntegral` a functor

Closes #29910

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2697958/), step Docs: sync website on a8f9d35 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29912/site)

This comment will be updated on new commits.

lindsayad commented:
AFAICT none of these failures are related

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2697955/) on a8f9d35 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2697974/), step Generate coverage on a8f9d35 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b05d63</th>
      <th colspan=4>#29912 a8f9d3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.27%</td>
      <td>-0.01%</td>
      <td>5.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108456</td>
      <td>108457</td>
      <td>+1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18711</td>
      <td>18728</td>
      <td>+17</td>
      <td>19</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29912/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29912/coverage/framework)


## Modules coverage

Mutiple base shas were found when retrieving head reports:
  220ffa759e0c35da505d31601f46879b44211504
  b05d63c35c7ca2c192a7177954d5cccf1c6dbe09

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29912/coverage/framework)
- `chemical_reactions` (FAILED)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29912/coverage/combined)
- `contact` (FAILED)
- `electromagnetics` (FAILED)
- `external_petsc_solver` (FAILED)
- `fluid_properties` (FAILED)
- `fsi` (FAILED)
- `functional_expansion_tools` (FAILED)
- `geochemistry` (FAILED)
- `heat_transfer` (FAILED)
- `level_set` (FAILED)
- `misc` (FAILED)
- `navier_stokes` (FAILED)
- `optimization` (FAILED)
- `peridynamics` (FAILED)
- `phase_field` (FAILED)
- `porous_flow` (FAILED)
- `ray_tracing` (FAILED)
- `rdg` (FAILED)
- `reactor` (FAILED)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29912/coverage/richards)
- `scalar_transport` (FAILED)
- `solid_mechanics` (FAILED)
- `solid_properties` (FAILED)
- `stochastic_tools` (FAILED)
- `thermal_hydraulics` (FAILED)
- `xfem` (FAILED)

</details>

## Warnings

- `framework` new line coverage rate 5.00% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2697974/), step Verify coverage on a8f9d35 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

lindsayad committed:
Add test

lindsayad committed:
Layered UO refactoring

This refactoring allows us to get layered element integral functor
quantities while DRY

GiudGiud commented:
why not make all the spatial user objects functors?

GiudGiud commented:
with a SpatialUserObjectFunctor<Real>  base class ?

lindsayad committed:
Create SpatialUserObjectFunctor per Guillaume's suggestion

lindsayad committed:
Fix layered side average error check

lindsayad committed:
Modify liquid fraction aux test since LineValueVPP is now a functor

lindsayad committed:
Dont pass Point by value

lindsayad committed:
Add thread 0 copy of functor UOs to functor database

This is because only the thread 0 copy has all the data required
for its functor evaluation

lindsayad committed:
Add .cache to .gitignore

GiudGiud commented on framework/include/userobjects/LayeredAverageBase.h:
```diff
@@ -0,0 +1,108 @@
+   */
+  virtual Real volume() const = 0;
+
+  /// Value of the volume for each layer
```
```suggestion
  /// Value of the volume (area for side integrals) for each layer
```

GiudGiud reviewed (changes requested):
Excited for this!

lindsayad commented on framework/include/userobjects/ElementIntegralFunctorUserObject.h:
```diff
@@ -0,0 +1,34 @@
+ * class and overriding computeQpIntegral().
+ */
+class ElementIntegralFunctorUserObject : public ElementIntegralUserObject,
+                                         public NonADFunctorInterface
```
I think we need a compelling case for using dense derivatives over PJFNK before changing this

GiudGiud commented on framework/include/userobjects/LayeredAverageFunctor.h:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Computes layered element averages of a functor.
+ */
+class LayeredAverageFunctor : public LayeredVolumeAverageBase<LayeredIntegralFunctor>
```
nevermind

GiudGiud replied:
> like both classes have the responsbility of performing the functor integral/sum ?

this isnt true, the layeredVolumeAverage only adds the volume averaging operation and the functor integral is handled by `LayeredVolumeIntegralBase<ElementIntegralFunctorUserObject`

GiudGiud replied:
ok this is fine. Just a bit advanced use of templating when inheritance would have done fine for adding  the volume adverage

GiudGiud replied:
but here inheritance would not maximize code reuse since every class introducing the averaging operation would need to have the same routines.

alternatively, I could see virtual inheritance work out with good code re-use. It s not the best either

lindsayad replied:
I cleaned up the hierarchy and put in a `hasBlocks` `override` in `SpatialUserObjectFunctor` that removes the need for the `LayeredVolumeIntegralBase` class

GiudGiud replied:
nice that's really neat

lindsayad committed:
Doxygen additions and fixes

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2710486/), step Clang format on a0daae5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29912/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29912/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 77637353649f8e0fa43943f62ce5e289388f6b97`

GiudGiud commented on framework/include/userobjects/LayeredAverageBase.h:
```diff
@@ -0,0 +1,108 @@
+#include "LayeredBase.h"
+
+/**
+ * Base class for computing layered side averages
```
```suggestion
 * Base class for computing layered averages
```

lindsayad committed:
cleanup class hierarchy per Guillaume's suggestion

lindsayad committed:
Add assertion that we are evaluating at the current state

lindsayad commented:
SAM patch at https://git-nse.egs.anl.gov/SAM/SAM/-/merge_requests/1020

lindsayad committed:
Add doc for new layered functor UOs

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2715315/) on 9dd1e30 : invalidated by @lindsayad

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2715327/) on 9dd1e30 : invalidated by @lindsayad

joshuahansel commented:
I was just thinking that I could use this for some work soon! This good to merge after next is opened up? 🏁

lindsayad commented:
Yea should be

GiudGiud commented:
@simopier this should give more options for coupling components of multiple dimensions together for mass transport

lindsayad commented:
I came up with this while working on PC's divertor monoblock inputs 👍
